src/hotspot/share/gc/g1/g1FullGCHeapRegionAttr.hpp

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_GC_G1_G1FULLGCHEAPREGIONATTR_HPP
+#define SHARE_GC_G1_G1FULLGCHEAPREGIONATTR_HPP
+
+#include "gc/g1/g1BiasedArray.hpp"
+
+// This table is used to store some per-region attributes needed during full
+// collection.
+class G1FullGCHeapRegionAttr : public G1BiasedMappedArray<uint8_t> {
+  static const uint8_t Normal = 0;
+  static const uint8_t Pinned = 1;
+  static const uint8_t ClosedArchive = 2;
+
+  static const uint8_t Invalid = 255;
```

Comment:
I do not understand what "Prior to this PR the default value is 0" means, can you clarify? The default value for the biased arrays must always be defined by the child classes overriding default_value() so there does not seem to be a pre-defined "prior" value.

Other than that the reason for using 255 is easier discrimination from the valid values in the debugger, just like hotspot uses something like 0xdeadbeef for various kinds of invalid memory.

There is no difference for the compiler for having 0 or 255 except for maybe the part about initializing the default value (on probably older x86 processors).

However compared to the overhead of setting up an efficient byte-wise `clear()` method is much larger than in this case a difference between `xor reg, reg` vs. `movsx reg, -1` (e.g. for x86, the former is/having been a common trick to quickly clear registers). That pales in comparison to the actual work.
Some architectures (ppc, sparc, iirc likely some more) have some special "clear cache line" instruction which may be used to clear memory "quickly" to zero, but it is not used since they often come with large caveats.

Otherwise the clearing of the this small table (one byte per region) is a one-time cost at startup of a full gc operations that may take billions of cpu cycles.

It's not worth thinking about overhead of the clearing in this case imho.

Reply:
> can you clarify?

In the original code, `uint8_t default_value() const { return NoArchive; }`, where `NoArchive = 0`.

>  It's not worth thinking about overhead of the clearing in this case imho.

OK, maybe it's not perf sensitive.